### PR TITLE
Add a warnings type and expose it usefully

### PR DESCRIPTION
**Pull request by rictic**

### CODE REVIEW

File: src/analyzer.ts [hunk 1]
```diff
@@ -203,16 +201,47 @@ export class Analyzer {
...
             };
-            const scannedDocument = await this._analyzeSource(
-                scannedDependency.type, scannedDependency.contents,
-                document.url, locationOffset,
-                scannedDependency.attachedComment);
-            scannedDependency.scannedDocument = scannedDocument;
-            scannedDependency.scannedDocument.isInline = true;
-            return scannedDocument;
+            try {
+              const scannedDocument = await this._analyzeSource(
```
Comment (justinfagnani) [inline src/analyzer.ts]: This function is getting to be too long. Can you split it up into focused subtasks?

Comment (justinfagnani) [inline src/analyzer.ts]: how about one method to analyze an `InlineParsedDocument` and another to analyze `ScannedImport`. The methods name themselves even.

File: src/analyzer.ts [hunk 2]
```diff
@@ -203,16 +201,47 @@ export class Analyzer {
...
+                code: 'could-not-load',
+                message: `Unable to load import: ${error.message || error}`,
+                sourceRange: scannedDependency.sourceRange,
+                severity: Severity.ERROR
+              });
+              return null;
+            }
             scannedDependency.scannedDocument = scannedDocument;
             return scannedDocument;
           } else {
```
Comment (justinfagnani) [inline src/analyzer.ts]: This is dead code. Maybe rewrite the filter/map so you don't feel the need for it?

File: src/analyzer.ts [hunk 3]
```diff
@@ -203,16 +201,47 @@ export class Analyzer {
...
-                scannedDependency.attachedComment);
-            scannedDependency.scannedDocument = scannedDocument;
-            scannedDependency.scannedDocument.isInline = true;
-            return scannedDocument;
+            try {
+              const scannedDocument = await this._analyzeSource(
+                  scannedDependency.type, scannedDependency.contents,
+                  document.url, locationOffset,
+                  scannedDependency.attachedComment);
+              scannedDependency.scannedDocument = scannedDocument;
```
Comment (justinfagnani) [inline src/analyzer.ts]: why is this method mutating these objects?

File: src/analyzer.ts (hunk 1 above)
Comment (rictic) [inline src/analyzer.ts]: +1, done

File: src/analyzer.ts (hunk 3 above)
Comment (rictic) [inline src/analyzer.ts]: Yeah it's super ugly. I didn't see another point in time where we know the mapping between the inline doc and its scanned document.

Setting whether a scanned document is inline could be done at a number of places though.

Thoughts?

File: src/analyzer.ts (hunk 2 above)
Comment (rictic) [inline src/analyzer.ts]: The above return is in a catch clause, this isn't dead.

The code is awkward to read though, pulled import and inline handling code into methods

Comment (rictic) [inline src/analyzer.ts]: Also renamed `analyze` to `scan` where appropriate`

Comment (rictic) [inline src/analyzer.ts]: IMO we should also rename `entityFinder` to `scanner`

**Comment by rictic** [comment]
Rebased onto master, only the most recent two commits are new to this PR

**Comment by justinfagnani** [comment]
LGTM

**Pull request merged**